File: src/typegraph/rust/typegraph/src/providers/prisma.rs
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::{
    t::{TypeBuilder, TypeDef},
    wasm::{
        self,
        core::RuntimeId,
        runtimes::{Effect, PrismaLinkData, PrismaRuntimeData},
        utils::TypeId,
    },
    Result,
};

#[derive(Debug)]
pub struct PrismaRuntime {
    id: RuntimeId,
    _name: String,
    _connection: String,
}

impl PrismaRuntime {
    pub fn new(name: &str, connection: &str) -> Result<Self> {
        let data = PrismaRuntimeData {
            name: name.to_string(),
            connection_string_secret: connection.to_string(),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_prisma_runtime(s, &data))?;

        Ok(Self {
            id,
            _name: data.name,
            _connection: data.connection_string_secret,
        })
    }

    pub fn find_unique<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_find_unique(s, self.id, model.into_id()?)).build()
    }

    pub fn find_many<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_find_many(s, self.id, model.into_id()?)).build()
    }

    pub fn find_first<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_find_first(s, self.id, model.into_id()?)).build()
    }

    pub fn aggregate<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_aggregate(s, self.id, model.into_id()?)).build()
    }

    pub fn count<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_count(s, self.id, model.into_id()?)).build()
    }

    pub fn group_by<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_group_by(s, self.id, model.into_id()?)).build()
    }

    pub fn create<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_create_one(s, self.id, model.into_id()?)).build()
    }

    pub fn create_many<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_create_many(s, self.id, model.into_id()?)).build()
    }

    pub fn update<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_update_one(s, self.id, model.into_id()?)).build()
    }

    pub fn update_many<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_update_many(s, self.id, model.into_id()?)).build()
    }

    pub fn delete<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_delete_one(s, self.id, model.into_id()?)).build()
    }

    pub fn delete_many<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_delete_many(s, self.id, model.into_id()?)).build()
    }

    pub fn upsert<T: TypeBuilder>(&self, model: T) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| r.call_prisma_upsert_one(s, self.id, model.into_id()?)).build()
    }

    pub fn execute<T: TypeBuilder>(
        &self,
        query: &str,
        param: T,
        effect: Effect,
    ) -> Result<TypeDef> {
        wasm::with_runtimes(|r, s| {
            r.call_prisma_execute(s, self.id, query, param.into_id()?, effect)
        })
        .build()
    }

    pub fn query_raw<P, O>(&self, query: &str, param: P, out: O) -> Result<TypeDef>
    where
        P: TypeBuilder,
        O: TypeBuilder,
    {
        wasm::with_runtimes(|r, s| {
            r.call_prisma_query_raw(s, self.id, query, Some(param.into_id()?), out.into_id()?)
        })
        .build()
    }

    pub fn link<T: TypeBuilder>(model: T) -> Result<PrismaLinkBuilder> {
        Ok(PrismaLinkBuilder {
            target: model.into_id()?,
            ..Default::default()
        })
    }
}

#[derive(Debug, Default, Clone)]
pub struct PrismaLinkBuilder {
    target: TypeId,
    name: Option<String>,
    fkey: Option<bool>,
    unique: Option<bool>,
    field: Option<String>,
}

impl PrismaLinkBuilder {
    pub fn name(mut self, name: &str) -> Self {
        self.name = Some(name.to_string());
        self
    }

    pub fn fkey(mut self) -> Self {
        self.fkey = Some(true);
        self
    }

    pub fn unique(mut self) -> Self {
        self.unique = Some(true);
        self
    }

    pub fn field(mut self, field: &str) -> Self {
        self.field = Some(field.to_string());
        self
    }
}

impl TypeBuilder for PrismaLinkBuilder {
    fn build(self) -> Result<TypeDef> {
        let data = PrismaLinkData {
            target_type: self.target,
            relationship_name: self.name,
            foreign_key: self.fkey,
            target_field: self.field,
            unique: self.unique,
        };

        wasm::with_runtimes(|r, s| r.call_prisma_link(s, &data)).build()
    }
}

File: src/typegraph/python/typegraph/providers/prisma.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typing import Union, Optional

from typegraph.runtimes.base import Runtime
from typegraph.sdk import runtimes
from typegraph.gen.runtimes import (
    Effect,
    PrismaRuntimeData,
    PrismaLinkData,
)
from typegraph import t
from typegraph.graph.typegraph import gen_ref


class PrismaRuntime(Runtime):
    name: str
    connection_string_secret: str

    def __init__(self, name: str, connection_string_secret: str):
        runtime_id = runtimes.register_prisma_runtime(
            PrismaRuntimeData(name, connection_string_secret)
        )

        super().__init__(runtime_id)
        self.name = name
        self.connection_string_secret = connection_string_secret

    def find_unique(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_find_unique(self.id, model._id)
        return t.func.from_type_func(type)

    def find_many(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_find_many(self.id, model._id)
        return t.func.from_type_func(type)

    def find_first(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_find_first(self.id, model._id)
        return t.func.from_type_func(type)

    def aggregate(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_aggregate(self.id, model._id)
        return t.func.from_type_func(type)

    def group_by(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_group_by(self.id, model._id)
        return t.func.from_type_func(type)

    def create(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_create_one(self.id, model._id)
        return t.func.from_type_func(type)

    def create_many(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_create_many(self.id, model._id)
        return t.func.from_type_func(type)

    def update(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_update_one(self.id, model._id)
        return t.func.from_type_func(type)

    def update_many(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_update_many(self.id, model._id)
        return t.func.from_type_func(type)

    def upsert(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_upsert_one(self.id, model._id)
        return t.func.from_type_func(type)

    def delete(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_delete_one(self.id, model._id)
        return t.func.from_type_func(type)

    def delete_many(self, model: Union[str, t.typedef]) -> t.func:
        if isinstance(model, str):
            model = gen_ref(model)
        type = runtimes.prisma_delete_many(self.id, model._id)
        return t.func.from_type_func(type)

    def execute(self, query: str, parameters: t.typedef, effect: Effect) -> t.func:
        type = runtimes.prisma_execute(self.id, query, parameters._id, effect)
        return t.func.from_type_func(type)

    def query_raw(
        self, query: str, parameters: Union[None, t.typedef], output: t.typedef
    ) -> t.func:
        params_id = None if parameters is None else parameters._id
        type = runtimes.prisma_query_raw(self.id, query, output._id, params_id)
        return t.func.from_type_func(type)

    def link(
        self,
        target_type: Union[str, t.typedef],
        name: Optional[str] = None,
        *,
        fkey: Optional[bool] = None,
        field: Optional[str] = None,
        unique: Optional[bool] = None,
    ):
        return prisma_link(
            target_type, name=name, fkey=fkey, field=field, unique=unique
        )


def prisma_link(
    target_type: Union[str, t.typedef],
    name: Optional[str] = None,
    *,
    fkey: Optional[bool] = None,
    field: Optional[str] = None,
    unique: Optional[bool] = None,
):
    if isinstance(target_type, str):
        target_type = gen_ref(target_type)
    type_id = runtimes.prisma_link(
        PrismaLinkData(
            target_type=target_type._id,
            relationship_name=name,
            foreign_key=fkey,
            target_field=field,
            unique=unique,
        ),
    )

    return t.typedef(type_id)
